### ISSUE TITLE
Unable to use dht2 function with a ds file that have rows without animals

### ISSUE THREAD

**Issue by jason-airst** [DistanceDevelopment/Distance#85]
When I try to use the dht2 function in distance using ds file I get the following error:
Error in predict.ds(ddf, newdata = bigdat, compute = TRUE) : fields or factor levels in `newdata` do not match data used in fitted model

The problem seems to be that the survey dataset has rows without animals. 
e.g.
Sample.Label	Effort	Region.Label	Area	object	size	distance	Species
40	3.675934875	Halifax	1056.55				
41	10	Halifax	1056.55				
42	9.653621341	Halifax	1056.55	81	2	0.002109644	D

When I removed these blank rows and run the function it  work. However, this gives faulty abundances and densities estimates. This is because the effort and cover area values used excludes areas without animals. `So,` both abundance and density are inflated.

**Comment by erex** [comment]
@jason-airst 

Jason

Thanks for the report, but I can't duplicate the error.  I used the `minke` data set that is included in the Distance package.  It has some characteristics in common with your data set: namely strata and transects without detections (9 of the 25 transects have no detections).  Running the code below, I was unable to generate an error

```
data("minke") # transects 5, 6, 9, 10, 11, 12, 17, 23, 25 without detections
minkefit <- ds(data=minke, transect = "line", key="hn")
minkedht2 <- dht2(ddf=minkefit, flatfile=minke, strat_formula = ~Region.Label, sample_fraction=.5)
```

Perhaps the problem has to do with a covariate in the detection function?  To check that, I manufactured a group size covariate in a new data file and tried that (see below)
```
whales <- read.csv("C:\\Users\\erexs\\Documents\\GitHub\\live-intro-7-stratification\\stratify_examp.csv")
wfit <- ds(data=whales, transect="line", key="hn", formula = ~size)
wdht2 <- dht2(ddf=wfit, flatfile = whales, strat_formula = ~Region.Label)
print(wdht2, report="abundance")
Summary statistics:
 Region.Label   Area CoveredArea Effort  n  k    ER se.ER cv.ER
        Ideal  85000     1877.92    484 39 13 0.081 0.019 0.232
     Marginal 600000     5315.60   1370 51 12 0.037 0.014 0.382
        Total 685000     7193.52   1854 90 25 0.049 0.024 0.484

Abundance estimates:
 Region.Label Estimate        se    cv   LCI   UCI     df group_mean  group_se
        Ideal     7903  2244.058 0.284  4457 14011 25.761   2.153846 0.2905123
     Marginal    26372 11129.370 0.422 11194 62127 16.317   2.411765 0.3477394
        Total    34275 11526.895 0.336 17017 69033 14.418   2.282805 0.4531226

Component percentages of variance:
 Region.Label Detection    ER Groups
        Ideal      8.23 68.64  23.13
     Marginal      3.74 84.25  12.01
        Total      2.31 83.64  14.05
```
again, no error message.

Please report the call to `ds()` as well as the call to `dht2()`.  That might help identify the problem.

As usual, also be sure you are using the most current version of the Distance package on CRAN (1.0.2).

**Comment by jason-airst** [comment]
Thank you for your quick response. I modified the "stratify_examp.csv" example you used by adding a species column for groups seen, so as to replicate my error. When I try to compare the detection curve for these species I get the error: _Error in predict.ds(ddf, newdata = bigdat, compute = TRUE) : fields or factor levels in newdata do not match data used in fitted model_. I am trying to replicate the example given in "Covariate modeling with rare species" on the distance sampling website. 

Here is the modified dataset:
Study.Area	Region.Label	Area	Sample.Label	Effort	distance	size	Species
Strat_Example	Ideal	85000	1	87	0.1	1	Minke
Strat_Example	Ideal	85000	1	87	0.22	1	Minke
Strat_Example	Ideal	85000	1	87	0.16	2	Minke
Strat_Example	Ideal	85000	1	87	0.78	4	Minke
Strat_Example	Ideal	85000	1	87	0.21	2	Minke
Strat_Example	Ideal	85000	1	87	0.95	6	Minke
Strat_Example	Ideal	85000	2	11	0	1	Minke
Strat_Example	Ideal	85000	2	11	0	3	Minke
Strat_Example	Ideal	85000	2	11	0.1	2	Minke
Strat_Example	Ideal	85000	2	11	0.06	5	Minke
Strat_Example	Ideal	85000	2	11	0.93	1	Minke
Strat_Example	Ideal	85000	2	11	0.38	2	Minke
Strat_Example	Ideal	85000	3	16	0.39	3	Minke
Strat_Example	Ideal	85000	3	16	0.61	2	Minke
Strat_Example	Ideal	85000	3	16	0.02	2	Minke
Strat_Example	Ideal	85000	4	110	0.34	1	Minke
Strat_Example	Ideal	85000	4	110	0.52	1	Minke
Strat_Example	Ideal	85000	4	110	0.36	2	Fin
Strat_Example	Ideal	85000	4	110	1.22	1	Fin
Strat_Example	Ideal	85000	4	110	0.68	2	Fin
Strat_Example	Ideal	85000	4	110	0.03	1	Fin
Strat_Example	Ideal	85000	4	110	0.26	1	Fin
Strat_Example	Ideal	85000	4	110	1.21	2	Fin
Strat_Example	Ideal	85000	4	110	0.52	2	Fin
Strat_Example	Ideal	85000	5	15
Strat_Example	Ideal	85000	6	10
Strat_Example	Ideal	85000	7	16	0.18	2	Fin
Strat_Example	Ideal	85000	7	16	0.36	2	Fin
Strat_Example	Ideal	85000	7	16	0.61	11	Fin
Strat_Example	Ideal	85000	7	16	0.61	1	Fin
Strat_Example	Ideal	85000	8	59	0.6	1	Fin
Strat_Example	Ideal	85000	8	59	0.1	2	Fin
Strat_Example	Ideal	85000	9	10
Strat_Example	Ideal	85000	10	13
Strat_Example	Ideal	85000	11	56
Strat_Example	Ideal	85000	12	1
Strat_Example	Ideal	85000	13	80	0.1	1	Minke
Strat_Example	Ideal	85000	13	80	0.68	1	Minke
Strat_Example	Ideal	85000	13	80	0.31	2	Minke
Strat_Example	Ideal	85000	13	80	0.58	2	Minke
Strat_Example	Ideal	85000	13	80	0.49	1	Minke
Strat_Example	Ideal	85000	13	80	0.46	2	Minke
Strat_Example	Ideal	85000	13	80	0.36	2	Minke
Strat_Example	Ideal	85000	13	80	0.09	2	Minke
Strat_Example	Ideal	85000	13	80	0.03	2	Minke
Strat_Example	Marginal	600000	14	75	0.49	1	Minke
Strat_Example	Marginal	600000	14	75	1.94	8	Minke
Strat_Example	Marginal	600000	14	75	1.1	10	Minke
Strat_Example	Marginal	600000	14	75	0.85	5	Minke
Strat_Example	Marginal	600000	14	75	0.63	7	Minke
Strat_Example	Marginal	600000	14	75	0.39	3	Minke
Strat_Example	Marginal	600000	14	75	0.65	1	Minke
Strat_Example	Marginal	600000	14	75	1.16	2	Minke
Strat_Example	Marginal	600000	14	75	0.91	2	Minke
Strat_Example	Marginal	600000	14	75	0	2	Minke
Strat_Example	Marginal	600000	14	75	1.18	10	Minke
Strat_Example	Marginal	600000	15	100	0.1	1	Minke
Strat_Example	Marginal	600000	15	100	0.37	1	Minke
Strat_Example	Marginal	600000	15	100	0.59	2	Minke
Strat_Example	Marginal	600000	15	100	0.45	2	Minke
Strat_Example	Marginal	600000	15	100	0.53	1	Minke
Strat_Example	Marginal	600000	16	110	0.21	2	Minke
Strat_Example	Marginal	600000	17	100
Strat_Example	Marginal	600000	18	125	0.85	2	Fin
Strat_Example	Marginal	600000	18	125	1.48	2	Fin
Strat_Example	Marginal	600000	18	125	1.44	2	Fin
Strat_Example	Marginal	600000	18	125	0.9	1	Fin
Strat_Example	Marginal	600000	18	125	0.01	1	Fin
Strat_Example	Marginal	600000	18	125	1.58	1	Fin
Strat_Example	Marginal	600000	18	125	0.68	1	Fin
Strat_Example	Marginal	600000	18	125	0.8	1	Fin
Strat_Example	Marginal	600000	18	125	0.64	1	Fin
Strat_Example	Marginal	600000	18	125	1.45	2	Fin
Strat_Example	Marginal	600000	18	125	0.31	3	Fin
Strat_Example	Marginal	600000	18	125	0.31	1	Fin
Strat_Example	Marginal	600000	18	125	0.85	1	Fin
Strat_Example	Marginal	600000	18	125	0.49	2	Fin
Strat_Example	Marginal	600000	18	125	0.67	3	Fin
Strat_Example	Marginal	600000	18	125	0.27	2	Fin
Strat_Example	Marginal	600000	18	125	0.56	1	Fin
Strat_Example	Marginal	600000	18	125	0.23	2	Fin
Strat_Example	Marginal	600000	19	100	0.39	1	Fin
Strat_Example	Marginal	600000	20	150	1.17	10	Fin
Strat_Example	Marginal	600000	20	150	0.97	1	Fin
Strat_Example	Marginal	600000	21	110	0.29	7	Fin
Strat_Example	Marginal	600000	21	110	0.59	2	Fin
Strat_Example	Marginal	600000	21	110	1.01	1	Fin
Strat_Example	Marginal	600000	21	110	0.87	1	Fin
Strat_Example	Marginal	600000	21	110	0.13	1	Fin
Strat_Example	Marginal	600000	21	110	0.78	1	Fin
Strat_Example	Marginal	600000	22	115	0.97	3	Fin
Strat_Example	Marginal	600000	22	115	0.48	1	Fin
Strat_Example	Marginal	600000	23	150
Strat_Example	Marginal	600000	24	125	0.41	1	Fin
Strat_Example	Marginal	600000	24	125	0.15	1	Fin
Strat_Example	Marginal	600000	24	125	0.09	1	Fin
Strat_Example	Marginal	600000	24	125	0.1	1	Fin
Strat_Example	Marginal	600000	24	125	0.1	1	Fin
Strat_Example	Marginal	600000	25	110

Here is the code I used:
wfit <- ds(data=Whale, transect="line", key="hn", formula = ~Species)
wdht2 <- dht2(ddf=wfit, flatfile = Whale, strat_formula = ~Species)
print(wdht2, report="abundance")

Hopefully you can reproduce the error now.

**Comment by erex** [comment]
Thanks for the additional information, yes I can reproduce the error.  I'll need to scratch my head about it.  The context of modelling detection functions by species gives some context for what you are trying to achieve.

I contrasted the fabricated 'minke' data set with the Montrave data set used in the "Covariate modeling with rare species" case study from the website.  Here's what I learned:

```
> table(birds$Sample.Label, birds$species)
    
      c  g  r  w
  1   4  0  3 10
  2   7  2  8 11
  3   7  3 11 12
  4   5  5  5 11
  5   9  5  8 14
  6   7  1  7 12
  7   5  3  5 13
  8   1  2  7 11
  9   1  1  3  6
  10  3  1  1  3
  11  1  0  0  1
  12  0  0  2  4
  13  2  2  5  9
  14  2  0  6 12
  15  4  2  4  9
  16  3  0  0  2
  17  4  3  4  7
  18  7  1  3  6
  19  1  1  0  3
> table(jason$Sample.Label, jason$Species)
    
        Fin Minke
  1   0   0     6
  2   0   0     6
  3   0   0     3
  4   0   7     2
  5   1   0     0
  6   1   0     0
  7   0   4     0
  8   0   2     0
  9   1   0     0
  10  1   0     0
  11  1   0     0
  12  1   0     0
  13  0   0     9
  14  0   0    11
  15  0   0     5
  16  0   0     1
  17  1   0     0
  18  0  18     0
  19  0   1     0
  20  0   2     0
  21  0   6     0
  22  0   2     0
  23  1   0     0
  24  0   5     0
  25  1   0     0
```
None of the transects for the bird data set were completely empty, there was a detection of at least one species on each transect.  Your 'minke' example differs in that some transects were derelict of detections of **any** species.   The absence of a "Species" entry for some transects set off the error message.

My "solution" (untested regarding whether it produces the correct answer) is to assign a "species" code to rows associated with empty transects.  I've done it programatically, but you could alter your raw data to include "Species" for your empty transects, making sure the perpendicular distance associated with this phantom detection is recorded as "NA".  Here's how I modified your 'minke' data:

```
jason <- read.csv("jason.txt", header=TRUE, sep=" ")
jason$Region.Label <- "singlestratum"
jason$Area <- 685000
jason$size <- NULL
jason$Species <- ifelse(jason$Sample.Label %in% c(5,6,9,10,11,12,17,23,25), "Minke", jason$Species)
wfit <- ds(data=jason, transect="line", key="hn", formula = ~Species)
wdht2 <- dht2(ddf=wfit, flatfile = jason, strat_formula = ~Species, stratification = "object")
print(wdht2, report="abundance")
```

and the result

```
Summary statistics:
 Species   Area CoveredArea Effort  n  k    ER se.ER cv.ER
     Fin 685000     7193.52   1854 47 25 0.025 0.010 0.394
   Minke 685000     7193.52   1854 43 25 0.023 0.009 0.396
   Total 685000     7193.52   1854 90 25 0.049 0.014 0.280

Abundance estimates:
 Species Estimate       se    cv   LCI   UCI     df
     Fin     9038 3734.216 0.413  4013 20358 28.959
   Minke    10126 4116.835 0.407  4536 22604 26.637
   Total    19164 5558.134 0.290 10842 33876 54.463

Component percentages of variance:
 Species Detection    ER
     Fin      3.65 96.35
   Minke      3.61 96.39
   Total      6.99 93.01
```
Note that the number of transects and the number of detections remains accurate (no new detections were manufactured, because there was no associated perpendicular distances for the 'phantom' detections).

Note also, I've knocked out the geographical strata that was present in the original minke data, because we have "commandeered" the "Region.Label" for species to produce species-specific estimates.  If geographical strata are important as well as needing the species-specific estimates, I think this approach to producing estimates for rare species might not work for the more complex situation that the one I described in the case study.

**Comment by jason-airst** [comment]
Thank you. Your solution does work.

**Comment by lenthomas** [comment]
Opening this back up as it seems like the workaround should not be necessary and there may be a bug in dht2.  Using species as a covariate in the detection function, and then also as a post-strat factor is a common way to get species-specific estimates, and it would also be reasonable to have no observations of any species on some transects.   Seems like post-strat factor needs to get passed to the predict function, and this isn't happening in cases where species doesn't exist on a transect.  @dill, can you take a look and self-assign if you agree it's an issue? Thanks!

**Comment by dill** [comment]
Fix in 0d27726

Answers now as above:

```
> print(wdht2, report="abundance")
Summary statistics:
 Species   Area CoveredArea Effort  n  k    ER se.ER cv.ER
     Fin 685000     7193.52   1854 47 25 0.025 0.010 0.394
   Minke 685000     7193.52   1854 43 25 0.023 0.009 0.396
   Total 685000     7193.52   1854 90 25 0.049 0.014 0.280

Abundance estimates:
 Species Estimate       se    cv   LCI   UCI     df
     Fin     9038 3734.216 0.413  4013 20358 28.959
   Minke    10126 4116.835 0.407  4536 22604 26.637
   Total    19164 5558.134 0.290 10842 33876 54.463

Component percentages of variance:
 Species Detection    ER
     Fin      3.65 96.35
   Minke      3.61 96.39
   Total      6.99 93.01
```
